### PR TITLE
`ParameterHandler::parse_input_from_json()`: do not use mangled parameter names anymore

### LINKED ISSUE #16453: ParameterHandler: *.json files with keys containing underscores cannot be parsed
# Problem description

In a student project, we are currently trying to parse an input file in `*.json` format with keys containing underscores. This is tested using the code example presented below, where we run into the following assert:

```bash
   terminate called after throwing an instance of 'dealii::ParameterHandler::ExcEntryUndeclared'
  what():  
--------------------------------------------------------
An error occurred in line <1234> of file </home/magdalena/code/external_libs/dealii/source/base/parameter_handler.cc> in function
    void dealii::ParameterHandler::set(const std::string&, const std::string&)
The violated condition was: 
    false
Additional information: 
    You can't ask for entry <test
Aborted (core dumped)
```

When we switch to `*.prm` format, everything works as expected. We assume that there is an inconsistency in mangling/demangling between boost and deal.II for `*.json` files (and maybe also for `*.xml` files (?)). At the moment I don't have a solution for a quick fix, but since I think this might be a common use case, I wanted to report this issue. We could at least add an appropriate assert. I will look into it asap.

@j-brotz FYI

# Test code 

```cpp
#include <deal.II/base/parameter_handler.h>
#include <fstream>
#include <iostream>

using namespace dealii;

int
main(int argc, char *argv[])
{
  ParameterHandler prm;

  double test_0 = 0;

 // test if underscore can be parsed
  prm.add_parameter("test_0", test_0);
  
  std::ifstream file;
  file.open(argv[1]);
  prm.parse_input_from_json(file);

  prm.print_parameters(std::cout, ParameterHandler::OutputStyle::ShortJSON);

  return 0;
}
```
tested with the file `test.json` 
```
{
 "test_0": "1"
}
```

### PR DESCRIPTION
**Pull request by mschreter**
As described in #16453 parameter names in `*.json` needed to be mangled. In this PR, we suggest to not use mangled parameter names in a `*.json`-file anymore. We mangle them internally to use the existing XML reading functionality. 
I also renamed the corresponding tests to keep the naming consistent, which was not the case before.

closes #16453

### CODE REVIEW

File: source/base/parameter_handler.cc [hunk 1]
```diff
@@ -355,10 +355,10 @@ namespace
...
 
-            recursively_demangle(p.second,
-                                 temp,
-                                 is_parameter_node(p.second) ||
-                                   is_alias_node(p.second));
+            recursively_mangle_or_demangle(p.second,
+                                           temp,
+                                           is_parameter_node(p.second) ||
+                                             is_alias_node(p.second));
             tree_out.put_child(demangle(p.first), temp);
```
Comment (marcfehling) [inline source/base/parameter_handler.cc:362]: So [`boost::core::demangle`](https://www.boost.org/doc/libs/1_84_0/libs/core/doc/html/core/demangle.html) works bidirectional from mangled to demangled and vice versa?

**Review by marcfehling — COMMENTED** [review]
After #13815, this PR is the only logical step to have a consistent input and output for the JSON format. Thank you.

The function no longer works for mangled JSON files, correct? I can imagine that this would be a big incompatibility for some users (including aspect? @gassmoeller @tjhei, referring to https://github.com/dealii/dealii/pull/13815#issuecomment-1146224331).

Is there a way to check whether a tree is mangled? Is there a way to convert a json file from mangled to demangled (that you could mention in the changelog)? This could be helpful for the transition.

**Comment by bangerth** [comment]
ASPECT doesn't use JSON. I think the compatibility issue is not huge. I'm ok with the incompatibility assuming it is properly documented in the changelog.

I had been under the impression that keys in JSON cannot have spaces in it, but looking at the changed output files, this seems to work just fine because both keys and values (if necessary) can be put in double quotes. So this is fine with me.

**Comment by bangerth** [comment]
@mschreter Ping?

**Comment by mschreter** [comment]
Thank you for your review @marcfehling.

> The function no longer works for mangled JSON files, correct? I can imagine that this would be a big incompatibility for some users (including aspect? @gassmoeller @tjhei, referring to [#13815 (comment)](https://github.com/dealii/dealii/pull/13815#issuecomment-1146224331)).

Correct, after this change json files are always assumed to be demangled. If this is really a big incompatibility, we probably need to extend this PR.

> Is there a way to check whether a tree is mangled? Is there a way to convert a json file from mangled to demangled (that you could mention in the changelog)? This could be helpful for the transition.

There is certainly a way to determine the state of the json file, but I am not sure if this is really what we want and I don't feel qualified enough to make this decision :-). However, if I interpret @bangerth's comment correctly, he would be fine with the incompatibility break in terms of requiring demangled json files in the future. Or do you @marcfehling/@bangerth have any specific use cases in mind where this PR could have a huge impact?

**Comment by tjhei** [comment]
@bangerth we use the JSON parameter output in ASPECT to generate the markdown documentation and this depends on mangling: https://github.com/geodynamics/aspect/blob/main/contrib/utilities/jsontomarkdown.py

This PR would break this. It wouldn't be hard to fix, but it is a bit annoying because we need to know what deal.II version was used.

**Comment by bangerth** [comment]
JSON does not allow for comments (one of the more startling decisions in the software world), so we cannot write the version number into a comment as we do in other file formats. We *could* put it into a "magic" variable, but that has the disadvantage that it is not backward compatible because previous version's readers do not know about the magicness of the variable.

**Comment by tjhei** [comment]
> JSON does not allow for comments (one of the more startling decisions in the software world),

True. It would also have the disadvantage that parsing the comment would be more difficult.

> We could put it into a "magic" variable, but that has the disadvantage that it is not backward compatible

Sadly we did not put all the output inside a "parameters" (or similar) block. Then we could include a version string. While breaking everybody, this would allow adding additional information...

**Comment by mschreter** [comment]
How shall we proceed @bangerth @tjhei? 

Another point which I would like to mention: The current implementation prints parameters in the *.json format in a demangled state. So we cannot read them anymore as long as we require mangled *.json files for parsing. I am afraid we have to fix this incompatibility one way or the other.

**Comment by mschreter** [comment]
ping :)

**Comment by peterrum** [comment]
> How do you feel about moving all entries into a single "parameters" root? That would allow us to add other information in the future (version info, etc.). You can even detect the two kinds of formats when loading (if there is a "parameters" root, you know you have the new version).

@tjhei  I am afraid that moving everything into a new section/node will break all json parameter files used by the users. This patch only fixes an inconsistency between reading/writing json files. At the moment we can write json files (by deal.II) in such a way that we cannot read it anymore.

**Comment by peterrum** [comment]
@mschreter Could you rebase the PR? I think after that we can merge :+1:

**Review by peterrum — APPROVED** [review]

**Comment by peterrum** [comment]
If there are no objections, I will merge this PR tomorrow.

**Pull request merged**